Feat/historic count

gowthamsundaresan opened the pull request:
Routes for historical count for avs, operator & staker based on 4 query params: `startAt`, `endAt`, `frequency`, `variant`

- Added 3 routes to `metricRoutes.ts`: `/historical/avs`, `/historical/operator` & `/historical/staker`
- in `metricController.ts`, setup respective functions for avs, operator & staker which call `doGetHistoricalCount()` passing the table name as a param
- `doGetHistoricalCount()` searches the db utilizing the `createdAt` column to get the tally & iterate over the range between `startAt` & `endAt` adding to its count every epoch defined in `frequency`

uditdc commented:
Looks great!

- For the timestamps in response, lets please reset the minutes and seconds to 00:00, currently set to the minutes and seconds of the request time.

- Another minor change, lets set the routes to `/metrics/historical/operators` and `/metrics/historical/stakers` as plural, to match other routes.

- For date range, there with the date range addition, there is no relative assignment of the `endDate` if a start date is set in the past, `endDate` becomes mandatory in that case, is that intentional ?

- For date range, we're able to set the `endDate` prior to the `startDate` without any validation warning, and the result is an empty array, also possible to replicate if just the `endDate` is set to a past date, the `startDate` doesn't relatively assign itself as it's set to 1 - today. 

Thanks!

gowthamsundaresan commented:
Hey thanks! Pushed updates and attaching notes against each point:
<br>

1. For the timestamps in response, lets please reset the minutes and seconds to 00:00, currently set to the minutes and seconds of the request time.
- Fix: Done, added `resetTime()` to `doGetHistoricalCount()`
```
/metrics/historical/stakers?startAt=2024-04-10T23:11:11.178Z&endAt=2024-04-15T23:13:21.342Z&frequency=1d&variant=cumulative
```
<br>

2. Another minor change, lets set the routes to `/metrics/historical/operators` and `/metrics/historical/stakers` as plural, to match other routes.
- Fix: Done, changed in `metricsRoutes.ts`
<br>

3. For the date range, with the date range addition, there is no relative assignment of the endAt if a start date is set in the past, endAt becomes mandatory in that case, is that intentional?
- Fix: !endAt && startAt
    - frequency === ‘1h’: endAt = startAt + 48h 
    - frequency === ‘1d’: endAt = startAt + 31d
    - frequency === ‘7d’: endAt = startAt + 52w
```
/metrics/historical/stakers?startAt=2024-03-10T21:00:10.178Z&frequency=1h&variant=cumulative
/metrics/historical/stakers?startAt=2024-03-10T21:00:10.178Z&frequency=1d&variant=cumulative
/metrics/historical/stakers?startAt=2024-03-10T21:00:10.178Z&frequency=7d&variant=cumulative
```
<br>

4. For the date range, we're able to set the endAt prior to the startAt without any validation warning, and the result is an empty array.
- Fix: Done, added validation error to historicalCountQuery.ts
```
/metrics/historical/stakers?startAt=2024-04-10T21:00:10.178Z&endAt=2024-03-07T21:00:10.178Z&frequency=7d&variant=cumulative
```
<br>

5. Also possible to replicate if just the endAt is set to a past date, the startAt doesn't relatively assign itself as it's set to 1 - today.
- Fix: endAt && !startAt
    - frequency === ‘1h’: startAt = endAt - 48h
    - frequency === ‘1d’: endAt = endAt - 31d
    - frequency === ‘7d’: endAt = endAt - 52w
    - Note: if !endAt && !startAt && !endAt, endAt is set to now and then the above gets executed
```
/metrics/historical/stakers?endAt=2024-03-07T21:00:10.178Z&frequency=1h&variant=cumulative
/metrics/historical/stakers?endAt=2024-03-07T21:00:10.178Z&frequency=1d&variant=cumulative 
/metrics/historical/stakers?endAt=2024-03-07T21:00:10.178Z&frequency=7d&variant=cumulative
```
<br>

6. For the date range, when selecting a different frequency (ie. 7d), the default startAt and endAt timeframe still follows the preset for 1h, thus resulting in a single point for a 7d frequency and two points for a 1d frequency. Ideally the default timeframe should take in consideration which frequency is selected.
- Fix: Default timeframes are set to the maximum return range for a given frequency (as demonstrated above)
```
/metrics/historical/stakers?frequency=1h&variant=cumulative
/metrics/historical/stakers?frequency=1d&variant=cumulative
/metrics/historical/stakers?frequency=7d&variant=cumulative
```

uditdc commented:
Looks good to me!
